### PR TITLE
Unnecessay cast / Multiloop / Hardware Screen / ControllerView

### PR DESCRIPTION
**Pull request by gi0e5b06**
1) The model is generic, the value can be of any type
2) Multiloop
3) Hardware-like widgets for emulating (long-term goal)
4) Make ControllerView similar to EffectView

### CODE REVIEW

**Comment by PhysSong** [comment]
I found there's other problems associated with type casting in the model.
Actually, `AutomatableModel::setValue()` gets `float` type value as an argument for any type of model.

**Comment by gi0e5b06** [comment]
The names are a bit strange and the design a little messy. But AutomatableModel is used to represents a CV (control voltage). So this is normal to use float, even for "IntModel" and "BoolModel". But refactoring is clearly needed.

**Comment by PhysSong** [comment]
@gi0e5b06 Welcome! You should make separate branches and add commits for separate issues.

**Comment by zonkmachine** [comment]
What ^he said.  Yeah, git takes some getting used to. That's for sure. I think the reason your PR looks a bit off now with 84 commits is that you merged your other features into this one branch. Start from master and make new branches.

Back up your master branch (this branch).
```
git checkout master
git checkout -b masterBACKUP
```

Then fix up master
 
Add an 'upstream' remote as suggested [here](https://github.com/LMMS/lmms/wiki/Accessing-git-repository) if you haven't done this yet.
```
git checkout master
git reset --hard HEAD~90  // backs master to some past good place
git merge upstream master  // if you've created a remote named upstream and 
```

Now your `master` will be synced to upstream/master.
This command should give no output:
`git diff master upstream/master`

Create a 'multiloop' branch from master and pick commits on top of it with the command `git cherry-pick`. These commits now live on the masterBACKUP branch and git will find them through the commit hash.
```
git checkout -b multiloop
git cherry-pick a0f9a994d36b81e364507ad393cd68e78bac76f4
git cherry-pick e5cc595a666ec1d4d869a56ef1a89344213cd1ba
```

Now you're set to push multiloop to your github and issue a pull request. Do the same for the invalid cast stuff. Create a new invalidcast branch and cherry-pick the commits manually like above. Then close this PR and open two new ones. One for each case. It's possible to keep this PR open but I think it will be less hassle to do it this way. I've also backed up this branch just in case...

**Comment by PhysSong** [comment]
@gi0e5b06 If you don't fix up this PR, we will close it. Please fix it, or request help to us.

File: include/TimeLineWidget.h [hunk 1]
```diff
@@ -123,12 +119,9 @@ class TimeLineWidget : public QWidget, public JournallingObject
 	inline QColor const & getActiveLoopInnerColor() const { return m_activeLoopInnerColor; }
 	inline void setActiveLoopInnerColor(QColor const & activeLoopInnerColor) { m_activeLoopInnerColor = activeLoopInnerColor; }
 
-<<<<<<< HEAD
 	inline QColor const & getActiveLoopTextColor() const { return m_activeLoopTextColor; }
 	inline void setActiveLoopTextColor(QColor const & activeLoopTextColor) { m_activeLoopTextColor = activeLoopTextColor; }
 
-=======
->>>>>>> refs/remotes/origin/master
```
Comment (zonkmachine) [inline include/TimeLineWidget.h]:
```diff
-=======
->>>>>>> refs/remotes/origin/master
```

This is what a merge gone wrong looks like. You should not merge anything. At least not for what you're trying to achieve. When you want to update a branch you `rebase`

Comment (PhysSong) [inline include/TimeLineWidget.h]: Once you do `git rebase`, you should use `git push -f` to force push.

File: plugins/CMakeLists.txt [hunk 2]
```diff
@@ -45,7 +45,9 @@ IF("${PLUGIN_LIST}" STREQUAL "")
 		carlabase
 		carlapatchbay
 		carlarack
+		ClickGDX
 		CrossoverEQ
+		csound_instr
```
Comment (zonkmachine) [inline plugins/CMakeLists.txt]: ClickGDX and csound_instr... ?

```
CMake Error at plugins/CMakeLists.txt:92 (ADD_SUBDIRECTORY):
  add_subdirectory given source "ClickGDX" which is not an existing
  directory.


CMake Error at plugins/CMakeLists.txt:92 (ADD_SUBDIRECTORY):
  add_subdirectory given source "csound_instr" which is not an existing
  directory.
```

File: include/QQ_MACROS.h [hunk 3]
```diff
@@ -38,7 +38,11 @@
 	public:    const type & getf() { return val; }\
 
 
+<<<<<<< a5433d807b1c19427640022fd1f38dedc6641a5f
```
Comment (zonkmachine) [inline include/QQ_MACROS.h]: If I get any odd messages I often search a file, or the whole code, for `<<<`.
It's a marker used by git on merge actions. You can see what files are failing with the command `git status`.

**Comment by gi0e5b06** [comment]
My Csound plugin (as an instrument) is only half-working. It plays the track, however the sound goes directly to ALSA, not to LMMS. So not really useful yet. I can commit it if someone can help working with the buffers.

ClickGDX is a small volume-based effect plugin. Not finished (I want to add a few params, stereo, etc) but it works.

QQ_MACROS is somehow an extended version of Q_PROPERTY, to avoid boilerplate code. Not used yet, have to decide first if it is useful.

**Comment by Umcaruje** [comment]
Let my just say that this will never get merged in it's current state. You need to break all of the changes into seperate Pull Requests that will be easy to review and merge. A huge PR like this is impossible to review and accept in the codebase. 

@zonkmachine wrote you a very handy guide on how to do this here, please follow it: https://gist.github.com/zonkmachine/e9b0fd99b61536f3db7ecfafa7118971

**Comment by zonkmachine** [comment]
@gi0e5b06 
 - Everything you put in your source directory is seen by git.
 - data/themes is part of the source directory.
 - If you put a new theme in data/themes it will be seen by git.
 - If you add a new theme in data/themes and do `git commit -am "fix"` after changing just one line of code, the new theme is going to go into that same commit.

You have pushed a new theme over the lmms default. Not just used a different theme locally but pushed a new theme in a PR for the LMMS team to review.

**Comment by PhysSong** [comment]
@gi0e5b06 File permissions are tracked unless you turn off the tracking(on Linux).

**Comment by gi0e5b06** [comment]
@PhysSong Well, I haven't met this option yet (only the ones to try to keep the formating) but it's a good thing anyway not to have executable image files. Every other data file seems to have the good attributes.

**Comment by zonkmachine** [comment]
> File permissions are tracked unless you turn off the tracking(on Linux).

Interesting, I didn't know file permissions where a thing in the git world!
In any case it shows clearly when testing the branch prior to a commit.

```
zonkmachine@zonkmachine:~/builds/lmms/lmms/data/themes/classic$ chmod +x zoom*

...

zonkmachine@zonkmachine:~/builds/lmms/lmms/data/themes/classic$ git diff
diff --git a/data/themes/classic/zoom.png b/data/themes/classic/zoom.png
old mode 100644
new mode 100755
diff --git a/data/themes/classic/zoom_x.png b/data/themes/classic/zoom_x.png
old mode 100644
new mode 100755
diff --git a/data/themes/classic/zoom_y.png b/data/themes/classic/zoom_y.png
old mode 100644
new mode 100755

...

zonkmachine@zonkmachine:~/builds/lmms/lmms/data/themes/classic$ git status 
HEAD detached at PhysSong/deadlock-fix
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   zoom.png
	modified:   zoom_x.png
	modified:   zoom_y.png
```

**Comment by zonkmachine** [comment]
> but it's a good thing anyway not to have executable image files

I agree. I'll open a separate issue for this and our graphic team should take a look at it.

**Comment by PhysSong** [comment]
Git has `core.filemode` setting. If it has value `true`, file mode change is tracked by Git.

**Comment by zonkmachine** [comment]
@gi0e5b06 What is the reason behind putting all your hacks into one pull request?
If you want to backup your stuff you just push it to your fork and don't open a PR. You will be prompted to open a PR but you can just ignore that. The message only pops up the first time you get on github.com/LMMS/lmms or your fork of lmms after pushing a new branch.

**Comment by PhysSong** [comment]
@gi0e5b06 We can do something with this PR, but it is very difficult because too many commits are mixed.
If you really want to contribute to LMMS, please make new branches for every feature/bug fix, sort your commits using `git rebase` or `git cherry-pick`, and open new PRs.

**Comment by zonkmachine** [comment]
> ClickGDX is a small volume-based effect plugin. Not finished (I want to add a few params, stereo, etc) but it works.

I tested this one. I actually tortured it under the new compile option `-DWANT_DEBUG_FPE=ON ` and it works. Precisely like the other commits it needs a separate pull request. Since it goes in the /plugin directory it's especially easy to break out from this PR. Please do this.

**Comment by zonkmachine** [comment]
@gi0e5b06 Here are some quotes form earlier in this thread.

>  Welcome! You should make separate branches and add commits for separate issues.

> You need to break all of the changes into seperate Pull Requests that will be easy to review and merg

And from the short manual I typed up to fix up this PR earlier:
> ... you merged your other features into this one branch. Start from master and make new branches.

...

The red thread in there is to make separate PR's for issues. Please do this.
Joint our developer channel at Discord  (linked earlier and from the LMMS/lmms front page).
You can also mail me if you have any questions.
Mail: `git show 020f1652bf08208458d53aa9d9fe038d1689b5d2 | grep Author`

**Comment by zonkmachine** [comment]
@gi0e5b06 Did you read my last post?

**Comment by gi0e5b06** [comment]
@zonkmachine The main problem is I need a working version with all my changes. Development is a global process. If you start opening branches for every change, development will stagnate (as you can see from the git history of lmms). Master should be the main branch for development but if you prefer two levels of stability, you can open a testing branch. So: extern -> master -> testing -> stable

Concerning ClickGDX, it's an effect that can be achieved easily by other means, is very light both for CPU and RAM and seems useful but maybe not for every body so being part of LMMS could be discussed.

By any standard, my PR is small. Someone should take 20 minutes to review it.

Concerning the last commits, it's the only way to test integration in the other environments. I'm hitting a strange behavior concerning qt properties that I don't understand. I'm following http://doc.qt.io/qt-5/properties.html but that doesn't compile everywhere. It compiles and runs fine locally with WANT_QT5=yes

**Comment by Spekular** [comment]
>By any standard, my PR is small.

The standard here is 1 PR per feature, so in terms of scope this is large.
It's far easier to judge wheter a PR works properly or not for a single
feature.

>Someone should take 20 minutes to review it.

That won't happen until its split. While LMMS is in need of devs, that
doesn't mean you can demand your PR gets reviewed, especially if you're not
following the same rules as everyone else. Reviews eat dev time too.

**Comment by zonkmachine** [comment]
> @zonkmachine The main problem is I need a working version with all my changes. Development is a global process. ... Master should be the main branch for development but if you prefer two levels of stability, you can open a testing branch. So: extern -> master -> testing -> stable

You can have a development branch on your own fork and run Travis there without pulling against `LMMS/lmms`.

> By any standard, my PR is small. Someone should take 20 minutes to review it.

It has already been reviewed. The outcome of this was a request to split the pull request into manageable parts. One PR per issue. That's the work flow here. Go with the flow...

**Comment by Umcaruje** [comment]
>By any standard, my PR is small. Someone should take 20 minutes to review it.

The problem is, even after reviewing, we don't have a reliable way of testing this because it touches so many files. LMMS has no unit tests, so accepting this mess would be a ticking time bomb for the codebase. 

This is also about having sustainable code, if you keep adding untested stuff, who's gonna fix your bugs when you're gone? We had this problem quite a few times, as people inevitably come and go on from the project.

You have some really nice ideas here and they would fit well with the rest of the software. It would be a shame to see your effort go to waste. 

This Pull Request won't ever be merged unless you cooperate with us and separate it into more manageable chunks that will get merged. If you want help or want to talk about this we are available over at our discord server: https://discord.gg/5kSc32Z

But please stop just adding stuff to the pull request and ignoring our requests, as @zonkmachine said this has been reviewed and the consensus is that this is unmergeable.

**Comment by lukas-w** [comment]
It's time to close this. This PR has been in an unmergeable state for long enough. @gi0e5b06 You've been ignoring all requests and all help offered to split this PR. It's become evident that you're not willing to cooperate. This is not something we can work with.

Feel free to open separate PRs for changes you feel should be included in LMMS. We appreciate any pull requests that are reviewable and mergeable by our standards and are happy to help in the process.

**Comment by gi0e5b06** [comment]
Please reopen so I can make sure my code compiles everywhere.
@PhysSong @zonkmachine @Umcaruje @Spekular @lukas-w @tresf @softrabbit

**Comment by tresf** [comment]
@gi0e5b06 please feel free to enable Travis-CI for your own repository, which will do the same thing.

https://docs.travis-ci.com/user/getting-started/

**Comment by Umcaruje** [comment]
This PR won't be opened unless you split it into manageable chunks.

**Pull request closed without merging**